File: frontend/src/components/timeline/timelineLayout.js
// Pure port of the iOS layout engine (ios/Beatably/Components/TimelineLayout.swift).
// Cards snake upward in rows of 3 (boustrophedon); the scale is computed from a
// fixed 4-row reference box so the board never rescales as cards accumulate.

export const NORMAL_SPACING = 100;
export const ROW_HEIGHT = 80;
export const MIN_MARGIN = 44;
export const CURVE_EXTEND = 58;

export const GAP_CIRCLE_SIZE = 24;
export const NODE_SIZE = 40;
export const NODE_CORNER_RADIUS = 16;
export const NODE_LABEL_OFFSET = 33;

// Returns { items, segments, scale, offsetX, offsetY }
//   items: { type: 'year', card, cardIndex, x, y } | { type: 'gap', index, x, y }
//   segments: { type: 'move'|'line', x, y } | { type: 'curve', x, y, c1x, c1y, c2x, c2y }
// `minMargin` lets a caller (e.g. the landing-page demo on a narrow phone)
// reclaim horizontal room by shrinking the board's side margin. Defaults to
// MIN_MARGIN so the game's own callers are unaffected.
export function calculateLayout(cards, containerSize, overrideOffsetY = null, minMargin = MIN_MARGIN) {
  const total = cards.length;

  // ── Step 1: raw positions ────────────────────────────────────
  const rawYears = [];
  for (let i = 0; i < total; i++) {
    const sectionIndex = Math.floor(i / 3);
    const posInSection = i % 3;
    const sectionY = -sectionIndex * ROW_HEIGHT;
    const isEven = sectionIndex % 2 === 0;
    const x = isEven ? posInSection * NORMAL_SPACING : (2 - posInSection) * NORMAL_SPACING;
    rawYears.push({ x, y: sectionY, sectionIndex });
  }

  // ── Step 2: bounding box of actual content ───────────────────
  const xs = rawYears.map((r) => r.x);
  const minX = (xs.length ? Math.min(...xs) : 0) - NORMAL_SPACING / 2;
  const maxX = (xs.length ? Math.max(...xs) : 0) + NORMAL_SPACING / 2;
  const rawW = maxX - minX;

  // ── Step 3: scale + offset (fixed 4-row reference box) ───────
  const fixedRawW = 3 * NORMAL_SPACING; // 300
  const fixedRawH = 3 * ROW_HEIGHT; // 240
  const availW = Math.max(containerSize.width - 2 * minMargin, 1);
  const availH = Math.max(containerSize.height - 2 * minMargin, 1);
  const scaleX = fixedRawW > availW ? availW / fixedRawW : 1;
  const scaleY = fixedRawH > availH ? availH / fixedRawH : 1;
  const scale = Math.min(scaleX, scaleY, 1);

  const scaledW = rawW * scale;
  const offsetX = containerSize.width / 2 - scaledW / 2 - minX * scale;
  const fixedMinY = -3 * ROW_HEIGHT; // -240
  const offsetY =
    overrideOffsetY != null
      ? overrideOffsetY
      : containerSize.height / 2 - (fixedRawH * scale) / 2 - fixedMinY * scale;

  const toScreen = (x, y) => ({ x: x * scale + offsetX, y: y * scale + offsetY });
  const years = rawYears.map((r) => ({ pos: toScreen(r.x, r.y), sectionIndex: r.sectionIndex }));

  // ── Step 4: path segments ─────────────────────────────────────
  const segments = [];
  const scaledCurve = CURVE_EXTEND * scale;

  if (total > 0) {
    const first = years[0];
    const firstGapX = first.pos.x - (NORMAL_SPACING / 2) * scale;
    segments.push({ type: 'move', x: firstGapX, y: first.pos.y });
    segments.push({ type: 'line', x: first.pos.x, y: first.pos.y });

    for (let i = 0; i < total - 1; i++) {
      const cur = years[i];
      const next = years[i + 1];
      const isVertical = Math.abs(next.pos.y - cur.pos.y) > 30 * scale;
      if (isVertical) {
        const isEvenSection = cur.sectionIndex % 2 === 0;
        const cx1x = cur.pos.x + (isEvenSection ? scaledCurve : -scaledCurve);
        const cx2x = next.pos.x + (isEvenSection ? scaledCurve : -scaledCurve);
        segments.push({
          type: 'curve',
          x: next.pos.x,
          y: next.pos.y,
          c1x: cx1x,
          c1y: cur.pos.y,
          c2x: cx2x,
          c2y: next.pos.y,
        });
      } else {
        segments.push({ type: 'line', x: next.pos.x, y: next.pos.y });
      }
    }

    const last = years[total - 1];
    const isEvenLast = last.sectionIndex % 2 === 0;
    const lastGapX = isEvenLast
      ? last.pos.x + (NORMAL_SPACING / 2) * scale
      : last.pos.x - (NORMAL_SPACING / 2) * scale;
    segments.push({ type: 'line', x: lastGapX, y: last.pos.y });
  }

  // ── Step 5: items (years interleaved with tappable gaps) ─────
  const items = [];
  let gapIdx = 0;

  if (total === 0) {
    items.push({
      type: 'gap',
      index: gapIdx,
      x: containerSize.width / 2,
      y: containerSize.height / 2,
    });
    return { items, segments, scale, offsetX, offsetY };
  }

  const first = years[0];
  items.push({
    type: 'gap',
    index: gapIdx++,
    x: first.pos.x - (NORMAL_SPACING / 2) * scale,
    y: first.pos.y,
  });

  for (let i = 0; i < total; i++) {
    const y = years[i];
    items.push({ type: 'year', card: cards[i], cardIndex: i, x: y.pos.x, y: y.pos.y });
    if (i < total - 1) {
      const next = years[i + 1];
      const isVertical = Math.abs(next.pos.y - y.pos.y) > 30 * scale;
      const midX = (y.pos.x + next.pos.x) / 2;
      const midY = (y.pos.y + next.pos.y) / 2;
      if (isVertical) {
        const isEvenSection = y.sectionIndex % 2 === 0;
        const shift = (ROW_HEIGHT / 2) * scale;
        items.push({
          type: 'gap',
          index: gapIdx++,
          x: isEvenSection ? midX + shift : midX - shift,
          y: midY,
        });
      } else {
        items.push({ type: 'gap', index: gapIdx++, x: midX, y: midY });
      }
    }
  }

  const last = years[total - 1];
  const isEvenLast = last.sectionIndex % 2 === 0;
  items.push({
    type: 'gap',
    index: gapIdx,
    x: isEvenLast
      ? last.pos.x + (NORMAL_SPACING / 2) * scale
      : last.pos.x - (NORMAL_SPACING / 2) * scale,
    y: last.pos.y,
  });

  return { items, segments, scale, offsetX, offsetY };
}

export function buildPathD(segments) {
  let d = '';
  for (const seg of segments) {
    if (seg.type === 'move') d += `M ${seg.x} ${seg.y}`;
    else if (seg.type === 'line') d += ` L ${seg.x} ${seg.y}`;
    else d += ` C ${seg.c1x} ${seg.c1y} ${seg.c2x} ${seg.c2y} ${seg.x} ${seg.y}`;
  }
  return d;
}

File: ios/Beatably/Components/TimelineLayout.swift
import CoreGraphics

// MARK: - Layout items

enum TimelineItemKind {
    case year(song: Song, cardIndex: Int)
    case gap(index: Int)
}

struct TimelineItem {
    let kind: TimelineItemKind
    let position: CGPoint
    var curveShift: CGFloat = 0  // perpendicular offset on curve segments

    var isYear: Bool { if case .year = kind { return true }; return false }
    var isGap: Bool  { if case .gap  = kind { return true }; return false }
    var gapIndex: Int? { if case .gap(let i) = kind { return i }; return nil }
    var song: Song? { if case .year(let s, _) = kind { return s }; return nil }

    var stableId: String {
        switch kind {
        case .year(let s, let cardIndex):
            // Include cardIndex so two copies of the same song (challenge-resolved:
            // originalCard + challengerCard) get distinct SwiftUI view identities.
            return "y-\(s.id)-\(cardIndex)"
        case .gap(let i):
            return "g-\(i)"
        }
    }
}

// MARK: - Path segment

enum PathSegment {
    case move(to: CGPoint)
    case line(to: CGPoint)
    case curve(to: CGPoint, control1: CGPoint, control2: CGPoint)
}

// MARK: - Result

struct TimelineLayoutResult {
    let items: [TimelineItem]
    let segments: [PathSegment]
    let scale: CGFloat
    let offsetX: CGFloat
    let offsetY: CGFloat
    // Height of the laid-out content. In scroll mode (solo) this can exceed the
    // viewport, and the container scrolls; otherwise it equals the viewport.
    var contentHeight: CGFloat = 0
}

// MARK: - Layout engine

enum TimelineLayout {
    static let normalSpacing: CGFloat = 100
    static let rowHeight: CGFloat = 80
    static let minMargin: CGFloat = 44
    static let curveExtend: CGFloat = 58

    static func calculate(
        cards: [Song],
        containerSize: CGSize,
        lastPlacedId: String?,
        gamePhase: String,
        isInteractive: Bool,
        overrideOffsetY: CGFloat? = nil,  // pin Y so adding rows doesn't shift existing content
        scrollMode: Bool = false          // solo: keep node size fixed and scroll instead of shrinking
    ) -> TimelineLayoutResult {

        // Lay out exactly the cards we're given. Callers (TimelineView.displayCards /
        // triggerAnimation) already apply phase-based filtering AND insert the pending card.
        // Filtering again here would drop the pending card during challenge, because its id
        // equals lastPlacedId — that collapsed the layout and made the placement + marker vanish.
        let confirmed = cards
        _ = (lastPlacedId, gamePhase)  // retained in signature for call-site clarity; not used for filtering

        let total = confirmed.count

        // ── Step 1: raw positions ────────────────────────────────────
        var rawYears: [(x: CGFloat, y: CGFloat, sectionIndex: Int, posInSection: Int)] = []
        for i in 0..<total {
            let sectionIndex = i / 3
            let posInSection = i % 3
            let sectionY = CGFloat(-sectionIndex) * rowHeight
            let isEven = sectionIndex % 2 == 0
            let x: CGFloat = isEven
                ? CGFloat(posInSection) * normalSpacing
                : CGFloat(2 - posInSection) * normalSpacing
            rawYears.append((x: x, y: sectionY, sectionIndex: sectionIndex, posInSection: posInSection))
        }

        // ── Step 2: bounding box ─────────────────────────────────────
        let minX = (rawYears.map(\.x).min() ?? 0) - normalSpacing / 2
        let maxX = (rawYears.map(\.x).max() ?? 0) + normalSpacing / 2
        let minY = rawYears.map(\.y).min() ?? 0
        let maxY = rawYears.map(\.y).max() ?? 0
        let rawW = maxX - minX
        let rawH = maxY - minY

        // ── Step 3: scale + offset ───────────────────────────────────
        // Scale is computed from the 4-row (12-card) bounding box so it never changes
        // as cards accumulate — the timeline stays visually stable all game.
        let fixedRawW: CGFloat = 3 * normalSpacing  // 300 — max section width
        let fixedRawH: CGFloat = 3 * rowHeight       // 240 — height spanning 4 sections
        let availW = max(containerSize.width  - 2 * minMargin, 1)
        let availH = max(containerSize.height - 2 * minMargin, 1)
        let scaleX = fixedRawW > availW ? availW / fixedRawW : 1
        let scaleY = fixedRawH > availH ? availH / fixedRawH : 1
        let scale  = min(scaleX, scaleY, 1)

        // Horizontal: centers actual content.
        let scaledW = rawW * scale
        let offsetX = containerSize.width  / 2 - scaledW / 2 - minX * scale

        // Vertical placement + content height. BOTTOM-anchor: the oldest row
        // (raw y = 0) sits `vpad` above the bottom of the board and rows grow
        // upward. This replaces vertical centering, which shifted the whole board
        // whenever the footer resized (album art appears during reveal). The
        // caller passes a stabilized height so a resizing footer doesn't move the
        // anchor. Solo overflow scrolls; everything else anchors to the bottom.
        let vpad = 40 * scale + 44   // node radius + year label
        let scaledContentH = rawH * scale
        let offsetY: CGFloat
        let contentHeight: CGFloat
        if scrollMode && scaledContentH + 2 * vpad > containerSize.height {
            contentHeight = scaledContentH + 2 * vpad
            offsetY = overrideOffsetY ?? (contentHeight - vpad)
        } else {
            contentHeight = containerSize.height
            offsetY = overrideOffsetY ?? (containerSize.height - vpad)
        }

        func toScreen(_ x: CGFloat, _ y: CGFloat) -> CGPoint {
            CGPoint(x: x * scale + offsetX, y: y * scale + offsetY)
        }

        let years = rawYears.map { r in
            (pos: toScreen(r.x, r.y), sectionIndex: r.sectionIndex, posInSection: r.posInSection)
        }

        // ── Step 4: path segments ─────────────────────────────────────
        var segments: [PathSegment] = []
        let scaledCurve = curveExtend * scale

        if total > 0 {
            let firstY = years[0]
            let firstGapX = firstY.pos.x - (normalSpacing / 2) * scale
            segments.append(.move(to: CGPoint(x: firstGapX, y: firstY.pos.y)))
            segments.append(.line(to: firstY.pos))

            for i in 0..<(total - 1) {
                let cur  = years[i]
                let next = years[i + 1]
                let isVertical = abs(next.pos.y - cur.pos.y) > 30 * scale
                if isVertical {
                    let isEvenSection = cur.sectionIndex % 2 == 0
                    let cx1x = cur.pos.x  + (isEvenSection ? scaledCurve : -scaledCurve)
                    let cx2x = next.pos.x + (isEvenSection ? scaledCurve : -scaledCurve)
                    segments.append(.curve(
                        to: next.pos,
                        control1: CGPoint(x: cx1x, y: cur.pos.y),
                        control2: CGPoint(x: cx2x, y: next.pos.y)
                    ))
                } else {
                    segments.append(.line(to: next.pos))
                }
            }

            let lastY = years[total - 1]
            let isEvenLast = lastY.sectionIndex % 2 == 0
            let lastGapX = isEvenLast
                ? lastY.pos.x + (normalSpacing / 2) * scale
                : lastY.pos.x - (normalSpacing / 2) * scale
            segments.append(.line(to: CGPoint(x: lastGapX, y: lastY.pos.y)))
        }

        // ── Step 5: items ─────────────────────────────────────────────
        var items: [TimelineItem] = []
        var gapIdx = 0

        if total == 0 {
            items.append(TimelineItem(kind: .gap(index: gapIdx),
                                      position: CGPoint(x: containerSize.width / 2, y: containerSize.height / 2)))
            return TimelineLayoutResult(items: items, segments: segments,
                                        scale: scale, offsetX: offsetX, offsetY: offsetY,
                                        contentHeight: contentHeight)
        }

        let firstY2 = years[0]
        let firstGapPos = CGPoint(x: firstY2.pos.x - (normalSpacing / 2) * scale, y: firstY2.pos.y)
        items.append(TimelineItem(kind: .gap(index: gapIdx), position: firstGapPos))
        gapIdx += 1

        for i in 0..<total {
            let y = years[i]
            items.append(TimelineItem(kind: .year(song: confirmed[i], cardIndex: i), position: y.pos))
            if i < total - 1 {
                let next = years[i + 1]
                let isVertical = abs(next.pos.y - y.pos.y) > 30 * scale
                let mid = CGPoint(x: (y.pos.x + next.pos.x) / 2, y: (y.pos.y + next.pos.y) / 2)
                if isVertical {
                    let isEvenSection = y.sectionIndex % 2 == 0
                    let shift = (rowHeight / 2) * scale
                    let shiftedX = isEvenSection ? mid.x + shift : mid.x - shift
                    var gap = TimelineItem(kind: .gap(index: gapIdx),
                                          position: CGPoint(x: shiftedX, y: mid.y))
                    gap.curveShift = isEvenSection ? shift : -shift
                    items.append(gap)
                } else {
                    items.append(TimelineItem(kind: .gap(index: gapIdx), position: mid))
                }
                gapIdx += 1
            }
        }

        let lastY2 = years[total - 1]
        let isEvenLast2 = lastY2.sectionIndex % 2 == 0
        let lastGapX2 = isEvenLast2
            ? lastY2.pos.x + (normalSpacing / 2) * scale
            : lastY2.pos.x - (normalSpacing / 2) * scale
        items.append(TimelineItem(kind: .gap(index: gapIdx),
                                  position: CGPoint(x: lastGapX2, y: lastY2.pos.y)))

        return TimelineLayoutResult(items: items, segments: segments,
                                    scale: scale, offsetX: offsetX, offsetY: offsetY,
                                    contentHeight: contentHeight)
    }
}
